[SPARK-8065] [hive] Add support for Hive 0.14 metastores

vanzin opened the pull request:
This change has two parts.

The first one gets rid of "ReflectionMagic". That worked well for the differences between 0.12 and
0.13, but breaks in 0.14, since some of the APIs that need to be used have primitive types. I could
not figure out a way to make that class work with primitive types. So instead I wrote some shims
 (I can already hear the collective sigh) that find the appropriate methods via reflection. This should
be faster since the method instances are cached, and the code is not much uglier than before,
with the advantage that all the ugliness is local to one file (instead of multiple switch statements on
the version being used scattered in ClientWrapper).

The second part is simple: add code to handle Hive 0.14. A few new methods had to be added
to the new shims.

Commit:
PART I: Get rid of reflection magic.

The reflection magic stuff in sql/hive does not work with primitive
Java types. Lots of APIs added in Hive 0.14 use primitive types in
their signature, so there's no way to make them work with that code.

Instead, have explicit shim-like classes that use reflection to call
into actual Hive code. Each version has its own version-specific
implementations of each needed method.

Commit:
Part II: Add shims / version for Hive 0.14.

Some dependency mess needed to be sorted out; in the process I removed the
unused constants for v10 and v11 (which, if desired, can be restored later
when someone decides to write shims for them).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34132/consoleFull) for   PR 6627 at commit [`a21cad8`](https://github.com/apache/spark/commit/a21cad8ede9d37cb93d765efc2c8c02c9fa1e3e8).

SparkQA commented:
[Test build #34132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34132/consoleFull) for   PR 6627 at commit [`a21cad8`](https://github.com/apache/spark/commit/a21cad8ede9d37cb93d765efc2c8c02c9fa1e3e8).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Style.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34133 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34133/consoleFull) for   PR 6627 at commit [`69b017b`](https://github.com/apache/spark/commit/69b017b4e40112bec4fc3b8753bcdc1c97c1c63d).

Commit:
Minor style thing.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34136 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34136/consoleFull) for   PR 6627 at commit [`6439d88`](https://github.com/apache/spark/commit/6439d88920ff25a88268ea43ba8c84e9822d46a4).

SparkQA commented:
[Test build #34133 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34133/consoleFull) for   PR 6627 at commit [`69b017b`](https://github.com/apache/spark/commit/69b017b4e40112bec4fc3b8753bcdc1c97c1c63d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #34136 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34136/consoleFull) for   PR 6627 at commit [`6439d88`](https://github.com/apache/spark/commit/6439d88920ff25a88268ea43ba8c84e9822d46a4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Fix NPE.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34195 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34195/consoleFull) for   PR 6627 at commit [`13b2dfa`](https://github.com/apache/spark/commit/13b2dfa2956f9eb96fdb40aaa4e11c854cbc825d).

SparkQA commented:
[Test build #34195 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34195/consoleFull) for   PR 6627 at commit [`13b2dfa`](https://github.com/apache/spark/commit/13b2dfa2956f9eb96fdb40aaa4e11c854cbc825d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

rxin commented:
I will let Michael comment on this. Explicitly not @ him so he won't see this during vacation; otherwise his heart might bleed... :)

vanzin commented:
The tests are not finding some dependencies that don't seem to exist in the main maven repo anymore but are available in other mirrors. Any suggestions for what to do here?

I can add another repo, or try to override the dependencies (although I'm not sure that will work since at least one has changed groupId).

vanzin commented:
Or I could remove the tests for 0.14 (and you'd have to trust me that they pass).

vanzin commented:
(Another minor update: the missing dependencies are all transitive dependencies of calcite. Not sure if it would be safe to upgrade the calcite version to a newer one. The same dependencies exist in Hive 1.0.0.)

Commit:
Try to fix tests by upgrading calcite dependency.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34215 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34215/consoleFull) for   PR 6627 at commit [`0a03470`](https://github.com/apache/spark/commit/0a0347033e2a5edd5dba86326cf5d1787c1649a5).

SparkQA commented:
[Test build #34215 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34215/consoleFull) for   PR 6627 at commit [`0a03470`](https://github.com/apache/spark/commit/0a0347033e2a5edd5dba86326cf5d1787c1649a5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

vanzin commented:
org.pentaho#pentaho-aggdesigner-algorithm is still missing after my last attempt... not sure what's the best way to proceed now (maven central doesn't have that artifact at all, not even a different version).

Commit:
Try to fix tests by adding support for exclusions.

Hive 0.14 has lots of dependencies that disappeared. Let's trim them
down so that at least the tests pass. Seems like anyone who'd like to
actually use it will have to do some Internet archaeology.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34252 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34252/consoleFull) for   PR 6627 at commit [`2c3c02e`](https://github.com/apache/spark/commit/2c3c02e3e24462a4b0febb47d497422ed9a760c6).

SparkQA commented:
[Test build #34252 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34252/consoleFull) for   PR 6627 at commit [`2c3c02e`](https://github.com/apache/spark/commit/2c3c02e3e24462a4b0febb47d497422ed9a760c6).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

Commit:
Merge branch 'master' into SPARK-8065

Conflicts:
	core/src/test/scala/org/apache/spark/deploy/SparkSubmitUtilsSuite.scala

Commit:
Comment about excluded dep.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34287 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34287/consoleFull) for   PR 6627 at commit [`d2ddf01`](https://github.com/apache/spark/commit/d2ddf01e825ff20264b70f5d7cbfe132d282b34b).

SparkQA commented:
[Test build #34287 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34287/console) for   PR 6627 at commit [`d2ddf01`](https://github.com/apache/spark/commit/d2ddf01e825ff20264b70f5d7cbfe132d282b34b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Botched merge.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34290 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34290/consoleFull) for   PR 6627 at commit [`73bd161`](https://github.com/apache/spark/commit/73bd16199d500d7cf9d8777bb32b9bd416e48ef3).

Commit:
Since exclusions seem to work, clean up some of the code.

SparkQA commented:
[Test build #34290 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34290/console) for   PR 6627 at commit [`73bd161`](https://github.com/apache/spark/commit/73bd16199d500d7cf9d8777bb32b9bd416e48ef3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

Commit:
Merge branch 'master' into SPARK-8065

Conflicts:
	core/src/test/scala/org/apache/spark/deploy/SparkSubmitUtilsSuite.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34307 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34307/consoleFull) for   PR 6627 at commit [`ca3fb1e`](https://github.com/apache/spark/commit/ca3fb1e4702aebb89b184038c865349d8e842d59).

SparkQA commented:
**[Test build #34307 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34307/console)**     for PR 6627 at commit [`ca3fb1e`](https://github.com/apache/spark/commit/ca3fb1e4702aebb89b184038c865349d8e842d59)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

WangTaoTheTonic commented:
A high level question: how about the hive 1.0/1.1/1.2? It might be hard to support so many versions if there's no compatibility between them.

vanzin commented:
I believe the API in 1.0/1.1/1.2 is very similar, if not identical, to 0.14. But this is a problem that we'll have to live with unless Spark is willing to reimplement some internal Hive APIs. While some of the metastore APIs used exist in public/stable form, at least a few (loadTable/loadPartitions) do not.

Commit:
Merge branch 'master' into SPARK-8065

Conflicts:
	core/src/test/scala/org/apache/spark/deploy/SparkSubmitUtilsSuite.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34453 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34453/consoleFull) for   PR 6627 at commit [`be3d0cc`](https://github.com/apache/spark/commit/be3d0cc1c2182828368e5b8b9c5edd7bf4873595).

SparkQA commented:
[Test build #34453 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34453/console) for   PR 6627 at commit [`be3d0cc`](https://github.com/apache/spark/commit/be3d0cc1c2182828368e5b8b9c5edd7bf4873595).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Fix dep exclusion.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34462 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34462/consoleFull) for   PR 6627 at commit [`4b8a3d4`](https://github.com/apache/spark/commit/4b8a3d428d246f81de91da048a0dc24c64734619).

SparkQA commented:
[Test build #34462 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34462/console) for   PR 6627 at commit [`4b8a3d4`](https://github.com/apache/spark/commit/4b8a3d428d246f81de91da048a0dc24c64734619).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

squito commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientWrapper.scala:
```diff
@@ -27,7 +27,7 @@ import scala.language.reflectiveCalls
 import org.apache.hadoop.hive.metastore.api.Database
 import org.apache.hadoop.hive.conf.HiveConf
-import org.apache.hadoop.hive.metastore.TableType
+import org.apache.hadoop.hive.metastore.{TableType => HTableType}
```
this change isn't required, is it?  just curious, I don't particularly care one way or the other

vanzin replied:
It gets rid of a scalac warning (since `TableType` is also defined in the Spark side).

squito commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/package.scala:
```diff
@@ -19,15 +19,27 @@ package org.apache.spark.sql.hive
+    //
+    // org.pentaho:pentaho-aggdesigner-algorithm is also nowhere to be found, so exclude
+    // it explicitly. If it's needed by the metastore client, users will have to dig it
+    // out of somewhere and use configuration to point Spark at the correct jars.
```
this comment is kinda scary ... do we need to document this somewhere?  though I'm not sure what the appropriate place would be

vanzin replied:
I'm sure there's gonna be some documentation needed; relying on Ivy to download these things is kind of a bad thing to start with (and I'd argue that it should be disabled by default).

In my view, the normal process would be for users to set the location of Hive jars in their configuration (along with the version to use) instead of relying on this. I think I saw an option for that but don't remember off the top of my head.

marmbrus replied:
By default Spark SQL will use the hive jars that are bundled with Spark, and for read deployments I expect most users will do as @vanzin suggests and set `spark.sql.hive.metastore.jars`.  This is really just for testing all the versions easily.  If someone wants to come up with a way to easily package all the various hive versions and find a way to get them on to jenkins / other places people want to run the tests we could use that instead.

@vanzin maybe we should open a JIRA for Hive though, as it would be good if we can depend on them directly eventually and the probably don't realize that is not possible.

vanzin replied:
> maybe we should open a JIRA for Hive though

Not sure how much they'd be able to fix; the dependencies that have disappeared are not Hive artifacts, and those Hive versions have already been published.

(Or maybe those have never been on maven central and you're supposed to use a 3rd-party repo to get them?)

marmbrus replied:
The JIRA I was thinking of would ask them to make sure that future versions of Hive can be used as a library simply by depending on them in maven.  Seems like they should be able to ensure future releases don't depend on SNAPSHOT dependencies and whatnot.

squito commented:
code changes lgtm.  I'm a little concerned about the dependency issues, but I'm not sure what to do about it

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -0,0 +1,267 @@
+  private lazy val setCurrentSessionStateMethod =
+    findStaticMethod(classOf[SessionState], "setCurrentSessionState", classOf[SessionState])
+  private lazy val setDataLocationMethod = findMethod(classOf[Table], "setDataLocation",
+    classOf[Path])
```
This is a nit, but the indenting style here is off and inconsistent with the above line (which would be my preference).  In SQL we usually use the following rules (which attempt to avoid having some arguments look 'special' due to indenting):
 - Break at the highest syntactic level (i.e. at the `=` not in the middle of a function argument list).
 - If you have to wrap arguments, wrap all arguments.

vanzin replied:
Changes it, hope I got the style right. Looks a bit noisy though.

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -0,0 +1,267 @@
+
+private[client] class Shim_v0_12 extends Shim {
+
+  private lazy val startMethod = findStaticMethod(classOf[SessionState], "start",
```
Minor, but do they actually have to be `lazy`?

vanzin replied:
Yes, because otherwise you can't have the shim for the next version extend this one. I mention this in the scaladoc for the `Shim` class.

marmbrus commented:
I'm okay with this change.  A few minor comments otherwise LGTM.

marmbrus commented:
Thanks for doing the upgrade BTW!

Commit:
Indentation style.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34735 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34735/consoleFull) for   PR 6627 at commit [`3fa4270`](https://github.com/apache/spark/commit/3fa42702a1060fd618baefc39344e32757a71d89).

SparkQA commented:
[Test build #34735 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34735/console) for   PR 6627 at commit [`3fa4270`](https://github.com/apache/spark/commit/3fa42702a1060fd618baefc39344e32757a71d89).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

marmbrus commented:
Thanks!  Merging to master.
